client/build_test.go

```diff
@@ -1069,6 +1073,339 @@ func testClientGatewayContainerPlatformPATH(t *testing.T, sb integration.Sandbox
 	checkAllReleasable(t, c, sb, true)
 }
 
+// testClientGatewayExecError is testing gateway exec to recreate the container
+// process for a failed execop.
+func testClientGatewayExecError(t *testing.T, sb integration.Sandbox) {
+	requiresLinux(t)
+
+	ctx := context.TODO()
+
+	c, err := New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	id := identity.NewID()
+	st := llb.Image("busybox:latest").Run(
+		llb.Dir("/src"),
+		llb.AddMount("/src", llb.Scratch()),
+		llb.Shlexf("sh -c \"echo %s > output && fail\"", id),
+	).Root()
+
+	def, err := st.Marshal(ctx)
+	require.NoError(t, err)
+
+	b := func(ctx context.Context, c client.Client) (*client.Result, error) {
+		_, solveErr := c.Solve(ctx, client.SolveRequest{
+			Evaluate:   true,
+			Definition: def.ToPB(),
+		})
+		require.Error(t, solveErr)
+
+		var se *errdefs.SolveError
+		require.True(t, errors.As(solveErr, &se))
+
+		op := se.Solve.Op
+		opExec, ok := se.Solve.Op.Op.(*pb.Op_Exec)
+		require.True(t, ok)
+
+		exec := opExec.Exec
+
+		var mounts []client.Mount
+		for _, mnt := range exec.Mounts {
+			mounts = append(mounts, client.Mount{
+				Selector:  mnt.Selector,
+				Dest:      mnt.Dest,
+				ResultID:  se.Solve.OutputIDs[mnt.Output],
```

Comment:
Both work, they could use `mnt.Input` but they would be the unmodified inputs. For this test, I was checking that I could access the mutated mounts rather than the original inputs.

In the LLB we have: `echo %s > output && fail`, so this test was to exec into a mount where `echo %s > output` succeeded but the `fail` failed the execop overall.

Reply:
The logic should be that both `se.Solve.OutputIDs` and `se.Solve.InputIDs` are indexed by `mnt.Input`. `mnt.Output` is a link to the next vertex. There is no need that output must be defined on a mount in order to be able to debug its error state.